django_emarsys/event.py

```diff
@@ -0,0 +1,263 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import logging
+
+from django.conf import settings
+from django.utils.html import conditional_escape
+
+from emarsys import EmarsysError
+
+from django_emarsys import api
+from django_emarsys.context_provider_registry import get_context_provider
+from django_emarsys.exceptions import (BadDataError, DjangoEmarsysError,
+                                       UnknownEventNameError)
+from django_emarsys.models import Event, EventInstance, EventParam
+
+log = logging.getLogger(__name__)
+
+
+def get_all_parameters_for_event(event_name):
+        return {argument: get_parameter_for_event(event_name, argument)
+                for argument in settings.EMARSYS_EVENTS[event_name].keys()}
+
+
+def get_parameter_for_event(event_name, argument):
+    name, model = settings.EMARSYS_EVENTS[event_name][argument]
+    return EventParam(
+        argument=argument,
+        name=name,
+        model=model
+    )
+
+
+def get_event_id(name):
+    try:
+        return Event.objects.get(name=name).emarsys_id
+    except Event.DoesNotExist:
+        pass
+
+    log.info("Emarsys-ID for event '{name}' unknown, "
+             "syncing...".format(name=name))
+
+    sync_events()
+
+    try:
+        return Event.objects.get(name=name).emarsys_id
+    except Event.DoesNotExist:
+        pass
+
+    return None
+
+
+def sync_events():
+    """
+    Sync locally available events, identified by their name, with emarsys.
+
+    In particular:
+        * Add events that don't exist locally.
+        * Update changed IDs.
+        * Delete Event objects if their name isn't reported by Emarsys
+          anymore. There are no foreign keys pointing to Event, so this
+          is safe.
+
+    Events are identified by their name. This allows adding and
+    implementing events in advance without knowing their emarsys id.
+    Syncing will then add a Event object with the emarsys ID so that they
+    can be triggered.
+
+    :return: num_new_events
+    """
+    num_new_events = 0
+    num_updated_ids = 0
+    num_deleted_ids = 0
+    unknown_event_names = []
+    unsynced_event_names = []
+
+    try:
+        emarsys_events = api.get_events()
+    except EmarsysError as e:
+        log.error('Emarsys error: {}\n'.format(e))
+        return num_new_events, num_updated_ids, num_deleted_ids, \
+            unknown_event_names, unsynced_event_names
```

Comment:
Hmmmmm. Any idea about the best way? The correct thing to do would be to re-raise the exception, I guess. But most of the calling code couldn't really make decisions on that anyway.